README.md
# ebayCollectibles

Contains lots of useful information on a list of vintage items that I am interested in collecting one day. 
I would like to come back to this project eventually and hook up a front end framework to it to better manage this information.

Analyzes vintage items and tracks their prices locally using eBay’s RESTful API. Provides access to all item statistics that can be retrieved at anytime without having to reissue more API requests through the usage of pickle files in Python.

•	Local logging and storage system that helps track and avoid errors

•	Batched item requests that allow flexible keyword searches across 20+ servers




logs/history.txt
##################################     SERVER: EBAY-US     ##################################
##################################     GAME: Game cube     ##################################
COUNT: 1 for keyword: Game cube -(cover, battery, tool, dvd, cable, gamepad, pad, hoodie, converter, adapter, usb, replacement) wata -(3.0, 3.5, 4.0, 4.5, 5.0, 5.5, 6.0, 6.5, 7.0, 7.5, 8.0, 8.5)
Duplicate Found VGA 85 WATA Luigis Mansion Luigi's factory sealed  nintendo GameCube game cube in server EBAY-US


COUNT: 2 for keyword: Game cube -(cover, battery, tool, dvd, cable, gamepad, pad, hoodie, converter, adapter, usb, replacement) vga -(50, 55, 60, 65, 70, 75, 80)
NEW GAME: Game cube
Game Cube GC Biohazard JAPAN RELEASE VGA 85 ARCHIVAL CASE
https://www.ebay.com/itm/Game-Cube-GC-Biohazard-JAPAN-RELEASE-VGA-85-ARCHIVAL-CASE-/301975170101




Duplicate Found VGA 85 WATA Luigis Mansion Luigi's factory sealed  nintendo GameCube game cube in server EBAY-US


NEXT SERVER


##################################     SERVER: EBAY-GB     ##################################
##################################     GAME: Game cube     ##################################
No results for server: EBAY-GB for Game cube

No results for server: EBAY-GB for Game cube

NEXT SERVER


##################################     SERVER: EBAY-AT     ##################################
##################################     GAME: Game cube     ##################################
No results for server: EBAY-AT for Game cube

COUNT: 2 for keyword: Game cube -(cover, battery, tool, dvd, cable, gamepad, pad, hoodie, converter, adapter, usb, replacement) vga -(50, 55, 60, 65, 70, 75, 80)
Duplicate Found Nintendo Game Cube , MARIO KART, DOUBLE DASH -  NEU- VGA 90 NM+/MT - GOLD - OVP in server EBAY-AT

NEW GAME: Game cube
Nintendo Game Cube ,THE LEGEND OF ZELDA , THE WIND WAKER - NEU - VGA 85 NM+
https://www.ebay.at/itm/Nintendo-Game-Cube-THE-LEGEND-OF-ZELDA-THE-WIND-WAKER-NEU-VGA-85-NM-/352742833975





NEXT SERVER


##################################     SERVER: EBAY-AU     ##################################
##################################     GAME: Game cube     ##################################
No results for server: EBAY-AU for Game cube

No results for server: EBAY-AU for Game cube

NEXT SERVER


##################################     SERVER: EBAY-CH     ##################################
##################################     GAME: Game cube     ##################################
No results for server: EBAY-CH for Game cube

COUNT: 3 for keyword: Game cube -(cover, battery, tool, dvd, cable, gamepad, pad, hoodie, converter, adapter, usb, replacement) vga -(50, 55, 60, 65, 70, 75, 80)
Duplicate Found Game Cube GC Biohazard JAPAN RELEASE VGA 85 ARCHIVAL CASE in server EBAY-CH

Duplicate Found Nintendo Game Cube , MARIO KART, DOUBLE DASH -  NEU- VGA 90 NM+/MT - GOLD - OVP in server EBAY-CH

Duplicate Found Nintendo Game Cube ,THE LEGEND OF ZELDA , THE WIND WAKER - NEU - VGA 85 NM+ in server EBAY-CH


NEXT SERVER


##################################     SERVER: EBAY-DE     ##################################
##################################     GAME: Game cube     ##################################
No results for server: EBAY-DE for Game cube

COUNT: 2 for keyword: Game cube -(cover, battery, tool, dvd, cable, gamepad, pad, hoodie, converter, adapter, usb, replacement) vga -(50, 55, 60, 65, 70, 75, 80)
Duplicate Found Nintendo Game Cube , MARIO KART, DOUBLE DASH -  NEU- VGA 90 NM+/MT - GOLD - OVP in server EBAY-DE

Duplicate Found Nintendo Game Cube ,THE LEGEND OF ZELDA , THE WIND WAKER - NEU - VGA 85 NM+ in server EBAY-DE


NEXT SERVER


##################################     SERVER: EBAY-ENCA     ##################################
##################################     GAME: Game cube     ##################################
COUNT: 1 for keyword: Game cube -(cover, battery, tool, dvd, cable, gamepad, pad, hoodie, converter, adapter, usb, replacement) wata -(3.0, 3.5, 4.0, 4.5, 5.0, 5.5, 6.0, 6.5, 7.0, 7.5, 8.0, 8.5)
Duplicate Found VGA 85 WATA Luigis Mansion Luigi's factory sealed  nintendo GameCube game cube in server EBAY-ENCA


COUNT: 2 for keyword: Game cube -(cover, battery, tool, dvd, cable, gamepad, pad, hoodie, converter, adapter, usb, replacement) vga -(50, 55, 60, 65, 70, 75, 80)
Duplicate Found Game Cube GC Biohazard JAPAN RELEASE VGA 85 ARCHIVAL CASE in server EBAY-ENCA

Duplicate Found VGA 85 WATA Luigis Mansion Luigi's factory sealed  nintendo GameCube game cube in server EBAY-ENCA


NEXT SERVER


##################################     SERVER: EBAY-ES     ##################################
##################################     GAME: Game cube     ##################################
No results for server: EBAY-ES for Game cube

COUNT: 3 for keyword: Game cube -(cover, battery, tool, dvd, cable, gamepad, pad, hoodie, converter, adapter, usb, replacement) vga -(50, 55, 60, 65, 70, 75, 80)
NEW GAME: Game cube
IKARUGA - VGA 90 NINTENDO GAMECUBE GAME CUBE PRECINTADO SEALED
https://www.ebay.es/itm/IKARUGA-VGA-90-NINTENDO-GAMECUBE-GAME-CUBE-PRECINTADO-SEALED-/164084750429




Duplicate Found Nintendo Game Cube , MARIO KART, DOUBLE DASH -  NEU- VGA 90 NM+/MT - GOLD - OVP in server EBAY-ES

Duplicate Found Nintendo Game Cube ,THE LEGEND OF ZELDA , THE WIND WAKER - NEU - VGA 85 NM+ in server EBAY-ES


NEXT SERVER


##################################     SERVER: EBAY-FR     ##################################
##################################     GAME: Game cube     ##################################
No results for server: EBAY-FR for Game cube

COUNT: 1 for keyword: Game cube -(cover, battery, tool, dvd, cable, gamepad, pad, hoodie, converter, adapter, usb, replacement) vga -(50, 55, 60, 65, 70, 75, 80)
Duplicate Found IKARUGA - VGA 90 NINTENDO GAMECUBE GAME CUBE PRECINTADO SEALED in server EBAY-FR


NEXT SERVER


##################################     SERVER: EBAY-FRBE     ##################################

tests/test_pickle.py
import pickle
from os.path import exists
import os
from collections import defaultdict
from pathlib import Path
from shutil import copy

# A with statement does not create a scope (like if, for, try/except and while do not create a scope either).

# Get real path to file and then append new directory name
new_items = os.path.dirname(os.path.realpath(__file__)) + "/" + "items"
new_logs = os.path.dirname(os.path.realpath(__file__)) + "/" + "logs"
# Make new directory
Path(new_items).mkdir(parents=True, exist_ok=True)
Path(new_logs).mkdir(parents=True, exist_ok=True)
# Directory of pickle file
dirof_item_results = new_items + "/" + "items.pkl"
dirof_history = new_logs + "/" + "history.txt"
dirof_id = new_logs + "/" + "item_id.txt"
dirof_log = new_logs + "/" + "log.txt"
# If file of items to read exists, then open it and set the list to equal what's inside the file, also make a backup
# Else create a new file and write a default dictionary of sets inside
item_results = defaultdict(set)
if exists(dirof_item_results):
    with open(dirof_item_results, "rb") as items_file:
        print("EXISTS")
        copy(dirof_item_results, new_items + "/" +"items_copy.pkl")
        items_results = pickle.load(items_file)
else:
    with open(dirof_item_results, "wb") as items_file:
        print("NOPE")
        pickle.dump(items_results, items_file, -1)



tests/test_exception.py
i = [0, 1, 2, 3, 4, 5]

# while(True):
#     continue
#     print "You won't see this"

# while(True):
#     pass
#     print "You will see this"
for x in i:
    try:
        if x == 3:
            raise AttributeError("ERROR")
    except AttributeError as e:
        print("we passed" + str(e))
    except Exception as e:
        print("HII")
    else:
        print(x)
      


tests/test_full_exception.py
from ebaysdk.finding import Connection as Finding
from ebaysdk.exception import ConnectionError

siteList = [
    "EBAY-US",
    "EBAY-GB",
    "EBAY-AT",
    "EBAY-AU",
    "EBAY-CH",
    "EBAY-DE",
    "EBAY-ENCA",
    "EBAY-ES",
    "EBAY-FR",
    "EBAY-FRBE",
    "EBAY-FRCA",
    "EBAY-HK",
    "EBAY-IE",
    "EBAY-IN",
    "EBAY-IT",
    "EBAY-MOTOR",
    "EBAY-MY",
    "EBAY-NL",
    "EBAY-NLBE",
    "EBAY-PH",
    "EBAY-PL",
    "EBAY-SG",
    "EBAY-US",
    "EBAY-GB",
    "EBAY-AT",
    "EBAY-AU",
    "EBAY-CH",
    "EBAY-DE",
    "EBAY-ENCA",
    "EBAY-ES",
    "EBAY-FR",
    "EBAY-FRBE",
    "EBAY-FRCA",
    "EBAY-HK",
    "EBAY-IE",
    "EBAY-IN",
    "EBAY-IT",
    "EBAY-MOTOR",
    "EBAY-MY",
    "EBAY-NL",
    "EBAY-NLBE",
    "EBAY-PH",
    "EBAY-PL",
    "EBAY-SG",
    "EBAY-US",
    "EBAY-GB",
    "EBAY-AT",
    "EBAY-AU",
    "EBAY-CH",
    "EBAY-DE",
    "EBAY-ENCA",
    "EBAY-ES",
    "EBAY-FR",
    "EBAY-FRBE",
    "EBAY-FRCA",
    "EBAY-HK",
    "EBAY-IE",
    "EBAY-IN",
    "EBAY-IT",
    "EBAY-MOTOR",
    "EBAY-MY",
    "EBAY-NL",
    "EBAY-NLBE",
    "EBAY-PH",
    "EBAY-PL",
    "EBAY-SG",
]

counter = 0
with open("test.txt", "w", encoding="utf-8") as test_file:
    for i in range(20):
        for site in siteList:
            try:
                api_request = {
                    "keywords": "Nintendo Wii",
                    "itemFilter": [{"name": "Condition", "value": "1000"}],
                }
                api = Finding(
                    domain="svcs.ebay.com",
                    config_file="ebay.yaml",
                    https=True,
                    siteid=site,
                )
                response = api.execute("findItemsAdvanced", api_request)
                if response.status_code == 404:
                    print("WE BROKE IN IF")
                    test_file.write("WE BROKE IN IF\n")
                else:
                    print(response.status_code)
                    test_file.write(str(response.status_code) + "\n")
                print("Finished server: " + str(counter))
                counter += 1
            except AttributeError as e:
                print(e)
                # print(e.response.dict())
                print("WE BROKE IN ATTRIBUTE ERROR")
                test_file.write("WE BROKE IN ATTRIBUTE\n")
                test_file.write(str(e) + "\n")
            except ConnectionError as e:
                print(e)
                print("WE BROKE IN CONNECTION ERROR")
                test_file.write("WE BROKE IN CONNECTION\n")
                test_file.write(str(e) + "\n")
            except Exception as e:
                print(e)
                print("WE SIMPLY BROKE")
                test_file.write("WE SIMPLY BROKE\n")
                test_file.write(str(e) + "\n")


parse_games.py
from collections import defaultdict
import itertools
import os

collection = os.path.dirname(os.path.realpath(__file__))
dirof_game_collection = collection + "/" + "game_collection.txt"
with open(dirof_game_collection, "r") as f:
    game_list = f.read().splitlines()


game_collection = defaultdict(list)
reach_system = False

# exclude keyword in searches
# gameList = ["Super Mario 64 -(DS, kart)", "Donkey Kong 64"]
for game in game_list:
    # After we parse system line, then we add system combinations
    game_split = game.split("-")
    if game == "[System]":
        reach_system = True
    elif reach_system:
        game_collection[game_split[0].strip()].append(
            game.strip() + " System" + " vga -(50, 55, 60, 65, 70, 75, 80)"
        )

        game_collection[game_split[0].strip()].append(
            game.strip()
            + " System"
            + " wata -(3.0, 3.5, 4.0, 4.5, 5.0, 5.5, 6.0, 6.5, 7.0, 7.5, 8.0, 8.5)"
        )
    else:
        game_collection[game_split[0].strip()].append(
            game.strip()
            + " wata -(3.0, 3.5, 4.0, 4.5, 5.0, 5.5, 6.0, 6.5, 7.0, 7.5, 8.0, 8.5)"
        )
        game_collection[game_split[0].strip()].append(
            game.strip() + " vga -(50, 55, 60, 65, 70, 75, 80)"
        )
# Strips should be good

# Game, itemURL, title, itemId in another file
# System with or without system

# Make chose number of combinations of each game and append to full list
# Change range based on number of combinations
# numCombinations = 1  ###################
# for index, subset in zip(
#         range(numCombinations), itertools.permutations(gameSplit, len(gameSplit))
#     ):


read_pickle.py
import pickle

from e_collector import Item
from collections import defaultdict


#75 VGA to Wata 90
#80 VGA to Wata 92
#85-90 or 9.4-9.6 and under is store bought
#85 to 9.4
#90 to 9.6
#95 to 9.8
#100 to 10.0
#wata over vga
#certified link and heritage auctions instead of ebay, ebay good for buying
#distributor channels
#circulars shopping nintendo games
curr = "items/new_items.pkl"
whole = "items/items.pkl"
with open(whole, "rb") as items_file:
    # full_list = defaultdict(set)
    # pickle.dump(full_list, items_file, -1)
    full_list = pickle.load(items_file)

# print(full_list["Pokemon Emerald"]) gives items
for z in full_list:
    # print(len(full_list[z]))
    #print(z)
    for item in full_list[z]:

        if item.game_name == "Super Mario 64":
            print(item.title)
            # print(item.item_url)
# any() function short-circuits and returns True as soon as a match has been found.
# if not any(x.item_id == i.item_id for x in itemList)

# for item in read_items('item_objects.pkl'):
#     for i in item:
#         print('  name: {}, value: {}'.format(i.site, i.title))

# write to new file new listings to parse
# for z in new_listings:
#   print("New Listing: " + str(z))

# item no results doesnt get anything in list



parsers/parse_new_item.py
from ebaysdk.shopping import Connection as Shopping
from ebaysdk.exception import ConnectionError
import json
from e_collector import Item
        # write to new file new listings to parse
        # for z in new_listings:
        #   print("New Listing: " + str(z))
# Finish with taking item set and finding all items with ids
try:
    api = Shopping(config_file="ebay.yaml")
    response = api.execute("GetSingleItem", {"ItemID": 202855239800})
    if response.reply.Ack == "Success":
        curr = Item(
            response.reply.Item.ConditionDisplayName,
            response.reply.Item.Country,
            "No Start Time",
            response.reply.Item.EndTime,
            "No Watch Count",
            response.reply.Item.Location,
            "Returns Accepted N/A",
            response.reply.Item.ConvertedCurrentPrice.value,
            response.reply.Item.ConvertedCurrentPrice._currencyID,
            response.reply.Item.ListingStatus,
            response.reply.Item.Title,
            response.reply.Item.ViewItemURLForNaturalSearch,
            response.reply.Item.GalleryURL,
            response.reply.Item.ItemID,
            "No Original Name",
            "No Server Name",
        )

except ConnectionError as e:
    print(e)
    print(e.response.dict())


e_collector.py
from ebaysdk.finding import Connection as Finding
from ebaysdk.exception import ConnectionError
from collections import defaultdict
from os.path import exists
from pathlib import Path
from shutil import copy
import json
import itertools
import pickle
import os
import parse_games

site_list = [
    "EBAY-US",
    "EBAY-GB",
    "EBAY-AT",
    "EBAY-AU",
    "EBAY-CH",
    "EBAY-DE",
    "EBAY-ENCA",
    "EBAY-ES",
    "EBAY-FR",
    "EBAY-FRBE",
    "EBAY-FRCA",
    "EBAY-HK",
    "EBAY-IE",
    "EBAY-IN",
    "EBAY-IT",
    "EBAY-MOTOR",
    "EBAY-MY",
    "EBAY-NL",
    "EBAY-NLBE",
    "EBAY-PH",
    "EBAY-PL",
    "EBAY-SG",
]
# site_list = ["EBAY-PL", "EBAY-SG"]
# site_list = ["EBAY-US"]
# EBAY-US	     United States
# EBAY-AT	     Austria
# EBAY-AU	     Australia
# EBAY-CH	     Switzerland
# EBAY-DE        Germany
# EBAY-ENCA	     Canada (English)
# EBAY-ES	     Spain
# EBAY-FR	     France
# EBAY-FRBE	     Belgium (French)
# EBAY-FRCA	     Canada (French)
# EBAY-GB	     UK
# EBAY-HK	     Hong Kong
# EBAY-IE	     Ireland
# EBAY-IN	     India
# EBAY-IT	     Italy
# EBAY-MOTOR	 US Motors
# EBAY-MY	     Malaysia
# EBAY-NL	     Netherlands
# EBAY-NLBE	     Belgium (Dutch)
# EBAY-PH	     Philippines
# EBAY-PL	     Poland
# EBAY-SG		 Singapore


class Item:
    def __init__(
        self,
        condition,
        country,
        startTime,
        endTime,
        watchCount,
        location,
        returns,
        price,
        currency,
        sellingState,
        title,
        item_url,
        image_url,
        item_id,
        game_name,
        site,
    ):
        self.condition = condition
        self.country = country
        self.startTime = startTime
        self.endTime = endTime
        self.watchCount = watchCount
        self.location = location
        self.returns = returns
        self.price = price
        self.currency = currency
        self.sellingState = sellingState
        self.title = title
        self.item_url = item_url
        self.image_url = image_url
        self.item_id = item_id
        self.game_name = game_name
        self.site = site

    def __eq__(self, other):
        return self.item_id == other.item_id

    def __hash__(self):
        return hash(self.item_id)

    def items_file(self):
        attr = vars(self)
        return "".join("%s: %s\n" % item for item in attr.items())


# Search through all words, maybe paralell
# If item expired, then remove from item_results and append new
# display image of each item with clickable url and price
# list comprehension


# 1000 is New, 1500 is New Other, 3000 is Used

# For quotes within quote: \"     \"
# Other keywords gold near mint sealed factory sealed

# for idx, item in enumerate(logList):
#      print(idx + 1)
#      print(item)


def check_attributes(item):
    for attr in [
        "country",
        "location",
        "returnsAccepted",
        "title",
        "galleryURL",
    ]:
        if not hasattr(item, attr):
            setattr(item, attr, "N/A")
    for attr in ["startTime", "endTime", "watchCount"]:
        if not hasattr(item.listingInfo, attr):
            setattr(item.listingInfo, attr, "N/A")
    for attr in ["value", "_currencyId"]:
        if not hasattr(item.sellingStatus.currentPrice, attr):
            setattr(item.listingInfo, attr, "N/A")

    if not hasattr(item.sellingStatus, "sellingState"):
        setattr(item.listingInfo, "sellingState", "N/A")
    if not hasattr(item.condition, "conditionDisplayName"):
        setattr(item.condition, "conditionDisplayName", "N/A")


# A = {10, 20, 30, 40, 80}
# B = {100, 30, 80, 40, 60}
# print (A.difference(B))
# print (B.difference(A))
# {10, 20}
# {100, 60}
# every game has a set
# Change w to a


def main():
    # Get real path to file and then append new directory name
    fold_items = os.path.dirname(os.path.realpath(__file__)) + "/" + "items"
    new_logs = os.path.dirname(os.path.realpath(__file__)) + "/" + "logs"
    # Make new directory
    Path(fold_items).mkdir(parents=True, exist_ok=True)
    Path(new_logs).mkdir(parents=True, exist_ok=True)
    # Directory of pickle file
    dirof_item_results = fold_items + "/" + "items.pkl"
    dirof_new_items = fold_items + "/" + "new_items.pkl"
    dirof_history = new_logs + "/" + "history.txt"
    dirof_id = new_logs + "/" + "item_id.txt"
    dirof_log = new_logs + "/" + "log.txt"
    with open(dirof_id, "a+", encoding="utf-8") as ids_file, open(
        dirof_log, "w", encoding="utf-8"
    ) as log_file, open(dirof_history, "w", encoding="utf-8") as history_file:
        # Set file to search from beginning to add to set
        ids_file.seek(0)
        ids_set = set(line.strip() for line in ids_file)

        temp_ids_set = set(ids_set)  ############################
        request_counter = 0

        # If file of items to read exists, then open it and set the list to equal what's inside the file, also make a backup
        # Else create a new file and write a default dictionary of sets inside
        curr_results = defaultdict(set)
        item_results = defaultdict(set)
        if exists(dirof_item_results):
            with open(dirof_new_items, "wb") as new_items_file:
                pickle.dump(curr_results, new_items_file, -1)
            with open(dirof_item_results, "rb") as items_file:
                copy(dirof_item_results, fold_items + "/" + "items_copy.pkl")
                try:
                    print("PICKLE LOADED")
                    item_results = pickle.load(items_file)
                except EOFError:
                    print("PICKLE EMPTY")
                    log_file.write("PICKLE EMPTY")
                    item_results = defaultdict(set)
        else:
            with open(dirof_item_results, "wb") as items_file:
                print("PICKLE DOESNT EXIST")
                pickle.dump(item_results, items_file, -1)

        for site in site_list:
            history_file.write(
                "##################################     SERVER: "
                + str(site)
                + "     ##################################\n"
            )

            for game in parse_games.game_collection.items():
                game_name, keyword_combo = game
                history_file.write(
                    "##################################     GAME: "
                    + str(game_name)
                    + "     ##################################\n"
                )
                for keyword in keyword_combo:
                    # "paginationInput": {"pageNumber": "1"},
                    # "sortOrder": "Best Match"
                    # Best Match Default
                    # PricePlusShippingHighest Sorts items by the combined cost of the item price plus the shipping cost, with highest combined price items listed first.
                    # PricePlusShippingLowest Sorts items by the combined cost of the item price plus the shipping cost, with the lowest combined price items listed first.
                    # WatchCountDecreaseSort Sorts items by watch count in decreasing order for the given site
                    # EndTimeSoonest Sorts items by end time, with items ending soonest listed first.
                    try:
                        api_request = {
                            "keywords": keyword,
                            "itemFilter": [{"name": "Condition", "value": "1000"}],
                        }
                        api = Finding(
                            domain="svcs.ebay.com",
                            config_file="ebay.yaml",
                            https=True,
                            siteid=site,
                        )
                        response = api.execute("findItemsAdvanced", api_request)
                        request_counter += 1
                    except ConnectionError as e:
                        print(e)
                        print(e.response.dict())
                        log_file.write("CONNECTION ERROR: " + str(e) + "\n")
                        log_file.write(str(e.response.dict()) + "\n\n")
                        # response = {"reply":{"ack"}}
                        # response.reply.ack = "fail"
                    except AttributeError as e:
                        print(e)
                        log_file.write("ATTRIBUTE ERROR: " + str(e) + " for keyword \n")
                        log_file.write(
                            str(keyword) + "\n" + "Server: " + str(site) + "\n\n"
                        )
                    except Exception as e:
                        print(e)
                        print("WE SIMPLY BROKE")
                        log_file.write("WE SIMPLY BROKE: " + str(e) + "\n\n")
                    # x = json.dumps(response.dict(), sort_keys=True, indent=4)
                    else:
                        if (
                            response.reply.ack == "Success"
                            and response.reply.searchResult._count != "0"
                        ):
                            history_file.write(
                                "COUNT: "
                                + str(response.reply.searchResult._count)
                                + " for keyword: "
                                + str(keyword)
                                + "\n"
                            )
                            for item in response.reply.searchResult.item:
                                # if title contains 7.0, blah blah dont add
                                if not hasattr(item, "itemId"):
                                    setattr(item, "itemId", "N/A")
                                if not hasattr(item, "title"):
                                    setattr(item, "title", "N/A")

                                # import black list from file
                                black_list = [
                                    " 3.0 ",
                                    " 3.5 ",
                                    " 4.0 ",
                                    " 4.5 ",
                                    " 5.0 ",
                                    " 5.5",
                                    " 6.0 ",
                                    " 6.5 ",
                                    " 7.0 ",
                                    " 7.5 ",
                                    " 8.0 ",
                                    " 8.5 ",
                                    " 50 ",
                                    " 55 ",
                                    " 60 ",
                                    " 65 ",
                                    " 70 ",
                                    " 75 ",
                                    " 80 ",
                                    " 80+ "
                                ]
                                # Add item to set if there is no duplicate ID and if title isn't blacklisted
                                if item.itemId not in ids_set:
                                    if not any(x in item.title for x in black_list):
                                        check_attributes(item)

                                        curr_item = Item(
                                            item.condition.conditionDisplayName,
                                            item.country,
                                            item.listingInfo.startTime,
                                            item.listingInfo.endTime,
                                            item.listingInfo.watchCount,
                                            item.location,
                                            item.returnsAccepted,
                                            item.sellingStatus.currentPrice.value,
                                            item.sellingStatus.currentPrice._currencyId,
                                            item.sellingStatus.sellingState,
                                            item.title,
                                            item.viewItemURL,
                                            item.galleryURL,
                                            item.itemId,
                                            game_name,
                                            site,
                                        )
                                        item_results[game_name].add(curr_item)
                                        curr_results[game_name].add(curr_item)
                                        # maybe not open and close everytime
                                        with open(dirof_item_results, "wb") as f:
                                            pickle.dump(
                                                item_results,
                                                f,
                                                protocol=pickle.HIGHEST_PROTOCOL,
                                            )
                                        with open(
                                            dirof_new_items, "wb"
                                        ) as new_items_file:
                                            pickle.dump(
                                                curr_results, new_items_file, -1
                                            )
                                        ids_set.add(item.itemId)
                                        ids_file.write(item.itemId + "\n")
                                        history_file.write(
                                            "NEW GAME: "
                                            + str(game_name)
                                            + "\n"
                                            + str(item.title)
                                            + "\n"
                                            + str(item.viewItemURL)
                                            + "\n\n\n\n\n"
                                        )
                                    else:
                                        history_file.write(
                                            "Black Listed Item: "
                                            + str(item.title)
                                            + " in server "
                                            + str(site)
                                            + "\n\n"
                                        )
                                else:
                                    history_file.write(
                                        "Duplicate Found "
                                        + str(item.title)
                                        + " in server "
                                        + str(site)
                                        + "\n\n"
                                    )
                            history_file.write("\n")
                        else:
                            history_file.write(
                                "No results for server: "
                                + str(site)
                                + " for "
                                + str(game_name)
                                + "\n\n"
                            )
                print(
                    "Processed Game: %s in Server: %s, API Counter: %d"
                    % (str(game_name), str(site), request_counter)
                )
            history_file.write("NEXT SERVER\n\n\n")
        history_file.write("DONE\n")
        new_items = ids_set.difference(temp_ids_set)
        history_file.write("NEW LISTINGS\n")
        history_file.write(", ".join(str(s) for s in new_items))
        history_file.write("\n")


if __name__ == "__main__":
    main()

